### ISSUE TITLE
Every Blazor project - the server responded with a status of 404 (Not Found)

### ISSUE THREAD

**Issue by troncomputers** [aspnet/AspNetCore#7788]
Hi!
Yesterday everything was fine. Today, every new/fresh project created with Visual Studio 2019 Preview is broken and it's showing only `Loading...` in browser.
Error in Google Chrome ver. 72.0.3626.109

```
Failed to load resource: the server responded with a status of 404 (Not Found)
components.webassembly.js:1 Uncaught (in promise) SyntaxError: Unexpected end of JSON input
    at Object.<anonymous> (components.webassembly.js:1)
    at components.webassembly.js:1
    at Object.next (components.webassembly.js:1)
    at i (components.webassembly.js:1)
```
It doesn't matter if project is clear `Blazor` or `Blazor (ASP.NET Core hosted)`. I didn't change anything. I just create a project and hit run on `IIS Express` button on top of VS.

```
dotnet --version
3.0.100-preview-010184
```
Blazor ver. 0.8.0
Latest ASP.NET Core Blazor Language Services

`dotnet build` - no errors
`dotnet run` - same behavior as build in VS.

PS.
I did add an issue in Blazor repo. Sorry about that. Can you remove it?
[aspnet/Blazor](https://github.com/aspnet/Blazor/issues/1790)

**Comment by troncomputers** [comment]
I did some kind of research what the problem could be and:
```javascript
Object.defineProperty(t, '__esModule', { value: !0 }),
      (t.fetchBootConfigAsync = function() {
        return r(this, void 0, void 0, function() {
          return o(this, function(e) {
            switch (e.label) {
              case 0:
                return [
                  4,
                  fetch('dist/_framework/blazor.boot.json', { // <= (Not Found) error. Here, I had to add 'dist'
                    method: 'Get',
                    credentials: 'include'
                  })
                ];
              case 1:
              return [2, e.sent().json()]; // <= Here is the second error.
            }
          });
        });
      }),
```
The `e.sent()` function returns HTML code which cannot be parsed to JSON.

**Comment by chucker** [comment]
A few things to check:

* you mean Visual Studio 2019 Preview. I assume this is preview **3**? You also have installed, separately, ASP.NET Core 3 preview **2**?

* make sure you don't have a stray `global.json` in the project in a recursive parent (grandparent, great-grandparent, …) directory that sets the wrong .NET Core runtime. IOW, try creating a fresh project in a completely different path and see if that helps.

* what does `dotnet --list-sdks` and `dotnet --list-runtimes` say? Do they mention `3.0.100-preview-010184` and `Microsoft.AspNetCore.App 3.0.0-preview-19075-0444`, respectively?

* is the base URL of this site `/`, or is there a path (like `/SomeProject`?)

**Comment by troncomputers** [comment]
- Yes, Visual Studio 2019 Preview 3 and ASP.NET Core 3 preview 2
- There is no `global.json` and same behavior if the project is created in different location, even on another partition
- `dotnet --list-sdks`
```
2.1.402 [C:\Program Files\dotnet\sdk]
2.1.403 [C:\Program Files\dotnet\sdk]
2.1.500 [C:\Program Files\dotnet\sdk]
2.1.502 [C:\Program Files\dotnet\sdk]
2.1.503 [C:\Program Files\dotnet\sdk]
2.1.504 [C:\Program Files\dotnet\sdk]
2.1.600-preview-009426 [C:\Program Files\dotnet\sdk]
2.1.600-preview-009472 [C:\Program Files\dotnet\sdk]
2.1.600-preview-009497 [C:\Program Files\dotnet\sdk]
2.2.100-preview3-009430 [C:\Program Files\dotnet\sdk]
2.2.101 [C:\Program Files\dotnet\sdk]
2.2.102 [C:\Program Files\dotnet\sdk]
3.0.100-preview-010184 [C:\Program Files\dotnet\sdk]
```
- `dotnet --list-runtimes`
```
Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.2.0-preview3-35497 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.All 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.2.0-preview3-35497 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.AspNetCore.App 3.0.0-preview-19075-0444 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.2.0-preview3-27014-02 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.2.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.NETCore.App 3.0.0-preview-27324-5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.WindowsDesktop.App 3.0.0-preview-27325-3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```
- there is no custom base URL. I didn't even publish this project. Every error is from Debug build with Visual Studio

**Comment by danroth27** [comment]
Does the project build cleanly?

Is the `dist/_framework/blazor.boot.json` file present on disk?

What is the content of the HTML response?

What do the ASP.NET Core logs say?

**Comment by troncomputers** [comment]
- Yes, everything build correctly
[debug_log.txt](https://github.com/aspnet/AspNetCore/files/2892561/debug_log.txt)
[build_log.txt](https://github.com/aspnet/AspNetCore/files/2892562/build_log.txt)
- No, there are only `.js` files 
```
dist/_framework/components.server.js
dist/_framework/components.webassembly.js
dist/_framework/wasm/mono.wasm
dist/_framework/wasm/mono.js
```
- content of HTML response is the `index.html` from `wwwroot`
```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width">
    <title>WebApplication1</title>
    <base href="/" />
    <link href="css/bootstrap/bootstrap.min.css" rel="stylesheet" />
    <link href="css/site.css" rel="stylesheet" />
</head>
<body>
    <app>Loading...</app>

    <script src="_framework/components.webassembly.js"></script>
</body>
</html>
```
- Would you please navigate me how to check ASP.NET Core logs? I'm not sure how to do that.

**Comment by chucker** [comment]
>* No, there are only `.js` files

So no `_bin` with DLLs inside either?

>* Would you please navigate me how to check ASP.NET Core logs? I'm not sure how to do that.

Navigate to the project (either the server project, or the _only_ project in a non-hosted template) in a command line, and typ `dotnet run`. This should build the project and run a web server. You should see output that ends with something like:

    Now listening on: http://localhost:64445
    Application started. Press Ctrl+C to shut down.

Now, navigate to that address in a browser. You should see requests, and possibly exceptions. I'm guessing Dan is looking for those.

**Comment by troncomputers** [comment]
- [Project screenshots](https://imgur.com/a/sTqG4BY) - screenshots from my project
- The result from `dotnet run`
```
info: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[0]
      User profile is available. Using 'C:\Users\Adam\AppData\Local\ASP.NET\DataProtection-Keys' as key repository and Windows DPAPI to encrypt keys at rest.
Hosting environment: Development
Content root path: D:\PROJEKTY\WebApplication1\WebApplication1.Server
Now listening on: http://localhost:54996
Application started. Press Ctrl+C to shut down.
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:54996/
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/index.html'. Physical path: 'D:\PROJEKTY\WebApplication1\WebApplication1.Client\wwwroot\index.html'
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 46.7053ms 200 text/html
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:54996/css/bootstrap/bootstrap.min.css
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:54996/css/site.css
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:54996/_framework/components.webassembly.js
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/css/site.css'. Physical path: 'D:\PROJEKTY\WebApplication1\WebApplication1.Client\wwwroot\css\site.css'
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 2.8075ms 200 text/css
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/css/bootstrap/bootstrap.min.css'. Physical path: 'D:\PROJEKTY\WebApplication1\WebApplication1.Client\wwwroot\css\bootstrap\bootstrap.min.css'
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 9.1984ms 200 text/css
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/_framework/components.webassembly.js'. Physical path: 'D:\PROJEKTY\WebApplication1\WebApplication1.Client\bin\Debug\netstandard2.0\dist\_framework\components.webassembly.js'
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 28.4301ms 200 application/javascript
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:54996/css/open-iconic/font/css/open-iconic-bootstrap.min.css
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/css/open-iconic/font/css/open-iconic-bootstrap.min.css'. Physical path: 'D:\PROJEKTY\WebApplication1\WebApplication1.Client\wwwroot\css\open-iconic\font\css\open-iconic-bootstrap.min.css'
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 4.5201ms 200 text/css
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:54996/_framework/blazor.boot.json
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 1.8736ms 404
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[1]
      Request starting HTTP/1.1 GET http://localhost:54996/favicon.ico
info: Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware[2]
      Sending file. Request path: '/index.html'. Physical path: 'D:\PROJEKTY\WebApplication1\WebApplication1.Client\wwwroot\index.html'
info: Microsoft.AspNetCore.Hosting.Internal.WebHost[2]
      Request finished in 0.475ms 200 text/html

```

**Comment by brianlowe** [comment]
I used the cli `dotnet publish`  to create my standalone app then set it up as a GitHub Pages site. The files are all present in the relevant branch (gh-pages) as required by GitHub *but* I got 404 for the request to `_framework/components.webassembly.js`.

I worked out that the underscore is the problem. Renaming the folder to `framework` (and making the corresponding edit in `index.html`) made GitHub send the .js file as expected. I had to also rename the `_bin` folder to `bin` and make corresponding edits for that too, before my GitHub site worked properly.

So, there is a workaround for getting a blazor site running on a service which doesn't like underscores in URL paths, but editing the generated .js files feels way too hacky.

I'd like to see a command-line argument or other config option to give the developer better control over the final published filesystem.

**Comment by brianlowe** [comment]
I loaded the same published resources (unhacked - retaining the underscores in URL paths) to a Windows 2008 server with IIS 8 and got 404 errors for the .json and .wasm files because there was no MIME mapping for those file types by default.

Adding custom MIME mappings for `.json' => `application/json` and `.wasm` => `application/wasm` resolved the issues.

**Comment by troncomputers** [comment]
- Is there a workaround for debugging with Visual Studio? 
- Have I always publish the project, change the names of the folders to see changes made in web app? 
- Will `.nojekyll` file work in debugging? Where to put it?

I've started Blazor project to learn the new technology. There is a big chance I will never publish that so do I need to wait for finall fix or you (@brianlowe) can tell me, step by step, how to get this working?

**Comment by brianlowe** [comment]
@troncomputers I'm also just at the experiment stage, so I just ran up the demo project in Visual Studio (2019 preview) but was frustrated when trying to run it under Visual Studio's IIS Express just failed. I find working in VS Code and at the command line a much nicer experience for DotNet Core.
Running it locally (at the command line with `dotnet run`) works fine (although I haven't done anything that needs debugging yet), but doesn't tell you anything about the problems you'll find when you need to deploy it.

Changing the folder names was a hack to get around the default GitHub gh-pages issue where it's server refuses to serve from a folder with a name that starts with an underscore. GitHub docs say you can put a file in the root of your gh-pages branch named `.nojekyll` to bypass that so your files will be served without you needing to change folder names. That's no help for debugging - just to get your blazor app served as a gh-pages site.

**Comment by troncomputers** [comment]
Ok, I can live with that. It is working with VS Code and `dotnet run` now, but the `dotnet watch` is not available for NET Core 3.0 so you won't see any changes unless you compile it again with `dotnet run`. I can't decide if I should wait for fix or just do it with VS Code and `dotnet run`... 🤔

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @troncomputers.
This is going to be addressed as part of #9014

**Comment by flgatormike** [comment]
I get this same error. I created a brand new project and it worked at first. Then I added some new packages from Nuget and then at some point it stopped working. I get this error:

Failed to load resource: the server responded with a status of 404 (Not Found)
http://localhost:64991/_framework/blazor.webassembly.js
